Fix Markov generator

N1ckP3rsl3y opened the pull request:
A few bugs have been found that deal with the markov weather generator which need to be fixed

- When the weather generator method is 2 (markov), dynamic memory is not allocated within `SW_RUN_deepCopy()` resulting in a possible segmentation fault or attempting to free a pointer that's already been freed (see #497)
- In SWNETCDF mode, we do not finalize the weather if weather is not read in (see #498)

N1ckP3rsl3y committed:
Allocate markov information before deep copying

fixes https://github.com/DrylandEcology/SOILWAT2/issues/497

- Now call `allocateMKV` before copying markov information to a new instance of SW_RUN

N1ckP3rsl3y committed:
Finalize weather with no weather generator in SWNETCDF mode

fixes https://github.com/DrylandEcology/SOILWAT2/issues/498

- `SW_NCIN_read_inputs()` now finalizes the weather after reading weather even if the weather generator is off

N1ckP3rsl3y committed:
Remove literals in weather generator comparisons

- Currently, when comparing weather generator values, we use literals, mainly the integer 2
- This is less readable than wgMKV

- Move wgLOCF (1), wgMKV (2) and new constant wgHist (0) to `SW_Weather.h`
- Use these constants wherever needed to eliminate the literals

N1ckP3rsl3y committed:
Update `NEWS.md` to mention changes of this branch

dschlaep commented on include/SW_Weather.h:
```diff
@@ -46,6 +46,16 @@
+static const unsigned int wgLOCF = 1;
+
+/** Markov weather generator method, see generateMissingWeather() */
+static const unsigned int wgMKV = 2;
```
- SW_MKV_setup() has two instances with `Weather_genWeathMethod == 2` that should now be `Weather_genWeathMethod == wgMKV`
- generateMissingWeather() has `method == 0` that should now be `method == wgHist`; it also has `method > 2` that should probably be `method != wgHist && method != wgLOCF && method != wgMKV`

dschlaep commented on src/SW_netCDF_Input.c:
```diff
@@ -8577,7 +8577,9 @@ void SW_NCIN_read_inputs(
+    if (runSims && readWeather) {
         for (input = 0; input < numInputs; input++) {
             SW_WTH_finalize_all_weather(
                 &sw->MarkovIn,
```
Doesn't a SW_NCIN_read_inputs() with `for (input ...) SW_WTH_finalize_all_weather(&sw->MarkovIn, ...) ` causes MarkovIn.markov_rng and maybe(?) markov_rng.ppt_events to be shared among the `numInputs`?

It seems we need a new struct SW_MARKOV_RUN_INPUTS that contains markov_rng and ppt_events which then becomes part of SW_RUN_INPUTS (and/or add SW_MARKOV_INPUTS to SW_RUN_INPUTS)? Then, this loop would become `for (input ...) SW_WTH_finalize_all_weather(inputs[input].MarkovRunIn, ...)`

N1ckP3rsl3y committed:
Update missed weather generator constants

- A previous commit (54b483a2d4fdfd32d862d0abd566bb0bfe60ebbd) introduced the use of existing/new weather generator constants instead of literals
- There are a couple uses of literals that should not be there

- Update `SW_MKV_setup()` & `generateMissingWeather()` to use "wgMKV" and "wgLOCF"

N1ckP3rsl3y committed:
Don't enter `generateMissingWeather()` generator method is wgHist (0)

- `generateMissingWeather()` is being entered when the weather generator method is 0 (wgHist) and then immediately returns
- This is a waste of a function call and would be better to simply skip the call to the function

- `finalizeAllWeather()` now checks if weather generator is > 0 before calling `generateMissingWeather()`

N1ckP3rsl3y committed:
Finalize weather before running a simulation

- We currently finalize weather in two places - `main()` (SWTXT or no input netCDF weather) or `SW_NCIN_read_inputs()`
- SW_MARKOV_INPUTS also must no longer be used across sites when finalizing weather

- Remove use of `SW_WTH_finalize_all_weather()` from `main()` and `SW_NCIN_read_inputs()`
- Now call `SW_WTH_finalize_all_weather()` before initializing the run in `SW_CTL_run_sw()`
- Update suid injection into `SW_CTL_run_sw()` to work with SWMPI and be seamless between modes

N1ckP3rsl3y committed:
Merge branch 'release/devel_v8.4.0' into feature_fixMarkovGenerator

N1ckP3rsl3y committed:
Fix creation of reference for `--mode=mpi`

fixes https://github.com/DrylandEcology/SOILWAT2/issues/500

- Set hardcoded --ntasks value to 2 in call to `Rscript__ncTestRuns_00_createReferenceRun.R`
- Set hardcoded log file name to "rank_0_logfile.log" from "logfile.log"

N1ckP3rsl3y committed:
Update `NEWS.md` to mention ncTestRun.sh fix

dschlaep committed:
ncTestRuns now checks the weather generator

- there is now more than one "reference" run and they are now organized in the new folder "results/referenceRuns/"
- the previous "results/referenceRun/" is moved and renamed to "results/referenceRuns/example/"
- a new reference run that utilizes the weather generator is "results/referenceRuns/example-wGen/"
- tools/check_ncTestRuns.sh option "-ref" now expects the path to the new folder that contains the individual reference runs (instead of the path to the output of the one reference run)
- "metadata_testRuns.csv" now contains a column "reference" which selects the reference run against which the test run is to be compared

- 3 new test runs that utilize the weather generator

dschlaep committed:
ncTestRuns now continuously print summary results instead of progress bar

dschlaep committed:
Update NEWS

N1ckP3rsl3y commented on tests/ncTestRuns/scripts/Rscript__ncTestRuns_01_createTestRuns.R:
```diff
@@ -403,16 +403,23 @@ pb <- utils::txtProgressBar(max = nrow(listTestRuns), style = 3L)
   #------ . ------
-  # Skip a testRun if no external weather data
-  if (isFALSE(hasWeather[[listTestRuns[k0, "inWeather"]]])) {
+  useWeatherGenerator <- identical(listTestRuns[k0, "inWeather"], "wGen")
```
When I run the command `./tools/check_ncTestRuns.sh cleanReference clean all`, upon the rest of the output, I get the following seemingly duplicated lines

```
   testRun-38     success       ok     values: ok            ok
   testRun-38     success       ok     values: ok            ok
   testRun-40     success       ok     values: ok            ok
   testRun-40     success       ok     values: ok            ok
```
where each line doesn't reuse an existing folder (i.e., two unique folders are created for test 38).

When I run `./tools/check_ncTestRuns.sh cleanReference clean all -t 39` I get the following error

```
Create ncTestRuns ...
Error in utils::txtProgressBar(max = nrow(listTestRuns), style = 3L) : 
  must have 'max' > 'min'
Execution halted

Execute and check ncTestRuns ...
Error: any(lengths(fnames_ref) > 0L) is not TRUE
Execution halted
```

Was this an intended modification?

dschlaep committed:
Fix identification numbers of ncTestRuns
